File: algorithms/Javascript/minNumCoinsForChange.js
function minNumOfCoinsForChange(n, denominations) {
	const numCoins = (new Array(n + 1)).fill(Infinity);
	numCoins[0] = 0;
	for (let denom of denominations) {
		for (let amount = 0; amount < numCoins.length; amount++) {
			if (denom <= amount) {
				numCoins[amount] = Math.min(numCoins[amount], numCoins[amount - denom] + 1);
			}
		}
	}
	return numCoins[n] !== Infinity ? numCoins[n] : -1;
}

File: algorithms/Python/minNumCoinsForChange.py
# O(nd) Time | O(n) Space
def minNumOfCoinsForChange(n, denominations):
    numCoins = [float("inf") for amount in range(n + 1)]
    # len n + 1 to include base case of 0 (change for $0)
    # infinity kickstarts algo as its always larger than any comparison of coin combos
    numCoins[0] = 0
    for denom in denominations:
        for amount in range(len(numCoins)): # range(len(n + 1))
            if denom <= amount:
                numCoins[amount] = min(numCoins[amount], 1 + numCoins[amount - denom])
    return numCoins[n] if numCoins[n] != float("inf") else -1